File: backend/app/api/routes/checkups.py
"""Admin checkup management API (JWT + RBAC: admin roles only)."""

from uuid import UUID

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from backend.app.auth.jwt import require_role
from backend.app.core.database import get_db
from backend.app.exceptions import ValidationError
from backend.app.models import HealthCheckReport, User
from backend.app.schemas.checkup import CheckupDetail, CheckupListItem, CheckupStats, CheckupSummary, IndicatorReviewOut, ReviewRequest
from backend.app.services.checkup_service import CheckupManagementService
from backend.app.services.health_report_service import parse_report_background

router = APIRouter(prefix="/checkups", tags=["admin-checkups"])
admin = require_role("admin", "company_admin", "system_admin")


@router.get("/summary", response_model=CheckupSummary)
async def checkup_summary(
    period: str = Query(default="30d", pattern="^(7d|30d|90d)$"),
    session: AsyncSession = Depends(get_db),
    _: User = Depends(admin),
):
    return await CheckupManagementService(session).summary(period)


@router.get("", response_model=list[CheckupListItem])
async def list_checkups(
    period: str = Query(default="30d", pattern="^(7d|30d|90d)$"),
    department: str | None = Query(default=None),
    status: str | None = Query(default=None),
    method: str | None = Query(default=None),
    session: AsyncSession = Depends(get_db),
    _: User = Depends(admin),
):
    items = await CheckupManagementService(session).list_reports(period=period, department=department, status=status, method=method)
    if status:
        items = [item for item in items if item["display_status"] == status or item["parse_status"] == status]
    return items


@router.get("/{report_id}", response_model=CheckupDetail)
async def checkup_detail(
    report_id: UUID,
    session: AsyncSession = Depends(get_db),
    _: User = Depends(admin),
):
    return await CheckupManagementService(session).get_detail(report_id)


@router.post("/{report_id}/reparse", response_model=CheckupDetail)
async def reparse_checkup(
    report_id: UUID,
    background_tasks: BackgroundTasks,
    session: AsyncSession = Depends(get_db),
    _: User = Depends(admin),
):
    """Re-run the existing parse pipeline on the original file (no re-upload)."""
    report = await session.get(HealthCheckReport, report_id)
    if report is None:
        raise HTTPException(404, "体检报告不存在")
    if report.parse_status == "parsing":
        raise HTTPException(400, "该报告正在解析中")
    report.parse_status = "uploaded"
    report.parse_progress = 5
    report.parse_error = None
    report.parsed_at = None
    await session.commit()
    await session.refresh(report)
    background_tasks.add_task(parse_report_background, report.id)
    return await CheckupManagementService(session).get_detail(report_id)


@router.patch("/indicators/{indicator_id}/review", response_model=IndicatorReviewOut)
async def review_checkup_indicator(
    indicator_id: UUID,
    payload: ReviewRequest,
    session: AsyncSession = Depends(get_db),
    current_user: User = Depends(admin),
):
    service = CheckupManagementService(session)
    try:
        indicator = await service.review_indicator(
            indicator_id, current_user.id,
            action=payload.action,
            reviewed_value=payload.reviewed_value,
            note=payload.note,
        )
        return IndicatorReviewOut(
            id=indicator.id, item_name=indicator.item_name, value=float(indicator.value),
            value_text=indicator.value_text, unit=indicator.unit,
            reference_text=indicator.reference_text, flag=indicator.flag,
            source_type=indicator.source_type,
            confidence=float(indicator.confidence) if indicator.confidence is not None else None,
            review_status=indicator.review_status,
            reviewed_value=float(indicator.reviewed_value) if indicator.reviewed_value is not None else None,
            review_note=indicator.review_note,
        )
    except ValidationError as exc:
        raise HTTPException(400, str(exc)) from exc

File: frontend/src/api/checkups.ts
import { request } from './request'

export interface CheckupStats { total_reports: number; parsed_count: number; pending_review_reports: number; abnormal_reports: number }
export interface CheckupListItem {
  id: string
  employee_name: string | null
  employee_no: string | null
  department: string | null
  report_name: string
  hospital: string | null
  report_date: string | null
  uploaded_at: string
  parse_method: string | null
  parse_mode: string | null
  indicator_count: number
  abnormal_count: number
  parse_status: string
  display_status: string
  pending_review: boolean
  ocr_used: boolean
  parse_error: string | null
}
export interface CheckupIndicator {
  id: string
  item_name: string
  value: number
  value_text: string | null
  unit: string | null
  reference_text: string | null
  flag: string
  source_type: string
  confidence: number | null
  review_status: string | null
  reviewed_value: number | null
  review_note: string | null
}
export interface CheckupDetail extends CheckupListItem {
  parse_warnings: Record<string, unknown> | null
  parsed_at: string | null
  items: CheckupIndicator[]
}
export interface DistributionItem { label: string; count: number; percent: number }
export interface CheckupSummary {
  stats: CheckupStats
  status_distribution: DistributionItem[]
  method_distribution: DistributionItem[]
  departments: string[]
}

export const getCheckupSummary = (period: string) => request.get<CheckupSummary>('/admin/checkups/summary', { params: { period } }).then(r => r.data)
export const getCheckups = (params: { period?: string; department?: string; status?: string; method?: string } = {}) =>
  request.get<CheckupListItem[]>('/admin/checkups', { params }).then(r => r.data)
export const getCheckup = (id: string) => request.get<CheckupDetail>(`/admin/checkups/${id}`).then(r => r.data)
export const reparseCheckup = (id: string) => request.post<CheckupDetail>(`/admin/checkups/${id}/reparse`).then(r => r.data)
export const reviewIndicator = (indicatorId: string, payload: { action: string; reviewed_value?: number; note?: string }) =>
  request.patch<CheckupIndicator>(`/admin/checkups/indicators/${indicatorId}/review`, payload).then(r => r.data)
